engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+
+	// serverconfig is a local copy of the serverCreateOpts struct generated
+	// with hcloud-go. This struct is dependent on the ServerType, Datacenter,
+	// Image and State params. These must be chosen from the valid options
+	// provided by Hetzner, see details on https://docs.hetzner.cloud/.
+	serverconfig hcloud.ServerCreateOpts
+
+	// rescueKeys is a local copy of the array of SSH key values to be enabled
+	// in rescue mode, after formatting for direct use with hcloud-go.
+	rescueKeys []*hcloud.SSHKey
+	// rescueEnable is an internal boolean flag to simplify the checkapply logic
+	// of operations related to rescue mode. False if ServerRescueMode is
+	// `disabled`, true for valid image types.
+	rescueEnable bool
+}
+
+// Default returns some safe defaults for this resource.
+func (obj *HetznerVMRes) Default() engine.Res {
+	return &HetznerVMRes{
+		WaitInterval:     HetznerWaitIntervalDefault,
+		WaitTimeout:      HetznerWaitTimeoutDefault,
+		ServerRescueMode: ServerRescueDisabled,
+		AllowDeletion:    "error",
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *HetznerVMRes) Validate() error {
+
+	// check for empty token
+	if obj.APIToken == "" {
+		return fmt.Errorf("empty token string")
+	}
+	// validate time params
+	if obj.MetaParams().Poll < HetznerPollLimit {
+		return fmt.Errorf("invalid polling interval (minimum %d s)", HetznerPollLimit)
+	}
+	if obj.WaitInterval < HetznerWaitIntervalLimit {
+		return fmt.Errorf("invalid wait interval (minimum %d)", HetznerWaitIntervalLimit)
+	}
+	// warning logs to inform user of interval limits
+	obj.init.Logf("WARNING: Meta:poll must always be greater or equal than %d seconds", HetznerPollLimit)
+	obj.init.Logf("WARNING: waitinterval must always be greater or equal than %d seconds", HetznerWaitIntervalLimit)
+
+	// validate state param
+	switch obj.State {
+	case StateRunning, StateOff, StateAbsent:
+		// valid options
+	case "":
+		return fmt.Errorf("invalid state: empty")
+	default:
+		return fmt.Errorf("invalid state: %s", obj.State)
+	}
+
+	// TODO: validate serverconfig params, cfr. https://docs.hetzner.cloud.
+	// These are prone to dynamic changes, consider API query at Init.
+	obj.init.Logf("WARNING: serverconfig options not explicitly validated")
+
+	// validate rescue mode parameters
+	switch obj.ServerRescueMode {
+	case ServerRescueTypeLinux32, ServerRescueTypeLinux64, ServerRescueTypeFreeBSD64:
+		// valid options for rescue mode image
+	case ServerRescueDisabled:
+		// valid option to disable rescue mode
+	default:
+		return fmt.Errorf("invalid serverrescuemode: %s", obj.ServerRescueMode)
+	}
+
+	return nil
+}
+
+// Init runs some startup code for this resource: initialize hcloud-go client,
+// and then build some internal flags from the given public fields.
+func (obj *HetznerVMRes) Init(init *engine.Init) error {
+
+	// save init struct
+	obj.init = init
+
+	// initialize hcloud-go client
+	obj.client = hcloud.NewClient(hcloud.WithToken(obj.APIToken))
```

Comment:
I'm adding those options, but I don't know how to access that `*lib.Main` object you refer to. From what I can tell it is related to the CLI operations, while I'm only interacting with the `*HetznerVMRes`. So far I can't really tell how to read those values, any tips would be appreciated. Otherwise I'll get back to it later.

Reply:
Fixed in commit `0ab2406db9b4d5dac2348dd033d5fd37abd92a94`. Use `obj.init.Program` and `obj.init.Version`.

Reply:
withEndpoint() uses `endpoint = "https://api.hetzner.cloud/v1"` by default, and I haven't found any information about valid alternatives yet. Not sure when it would be relevant to specify this as a resource param. Ideas?

withDebugWriter is currently added as TODO, will try to implement that in a later PR.